### PR TITLE
CLIENTS-773: Bar layout type for Contribution offer

### PR DESCRIPTION
**Pull request by tomasstark**
This PR addresses [CLIENTS-773](https://laterpay.atlassian.net/browse/CLIENTS-773).

- Adds template and styles for new **Bar** contribution type.

- Adjusts existing code for custom contribution amount to be compatible with AMP.

- Refactors [frontend script for contributions](https://github.com/laterpay/revenue-generator/blob/CLIENTS-773-contribution-bar/assets/src/js/frontend/contribution.js) to vanilla JS to remove jQuery dependency.

- Changes how custom contributions are processed. Offloads most of the logic to [the backend](https://github.com/laterpay/revenue-generator/blob/CLIENTS-773-contribution-bar/inc/classes/class-frontend-post.php#L726). Instead of checking for the amount entered by a user and then creating dedicated contribution URL on frontend, we submit data to backend using ajax and [validate everything server-side](https://github.com/laterpay/revenue-generator/blob/CLIENTS-773-contribution-bar/inc/classes/class-frontend-post.php#L729). The backend then responds with the URL that JS uses in `window.open` to open that URL in new tab. If form is submitted from AMP enabled page, [we pass headers for AMP](https://github.com/laterpay/revenue-generator/blob/CLIENTS-773-contribution-bar/inc/classes/class-frontend-post.php#L786) to handle the redirect the proper way. Moving this to server is also helpful for when we make contributions powered by tapper.

- Adds helper methods to easily [get currency](https://github.com/laterpay/revenue-generator/blob/CLIENTS-773-contribution-bar/inc/classes/class-client-account.php#L343) set by merchant, and [endpoints](https://github.com/laterpay/revenue-generator/blob/CLIENTS-773-contribution-bar/inc/classes/class-client-account.php#L324) based on locale and live mode setting.

- Reduces code duplication now that more templates are introduced to contributions. Moved common parts to individual files. Note files with `partial-` prefix in [templates folder](https://github.com/laterpay/revenue-generator/tree/CLIENTS-773-contribution-bar/templates/frontend/contribution).

### CODE REVIEW

File: inc/classes/class-client-account.php [hunk 1]
```diff
@@ -299,4 +316,41 @@ private function get_api_fetch_url() {
...
+
+		return $endpoints;
+	}
+
+	/**
+	 * Get currency from region.
+	 *
+	 * @return array Currency code and symbol.
+	 */
+	public function get_currency() {
```
Comment (fgenghini) [inline inc/classes/class-client-account.php:343]: I would change this function to use the same approach that you used on `web_endpoints`: create an array with the currency details, indexed by region. Then I would return:
```
return $currency_details[ $this->merchant_region ]
``` 
Doing so, if we have a new currency later on, we can simply add a new item to the array and we won't need to deal with the conditionals, following the [open/close principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) (of course, if we have a new currency, plenty of changes will be required in the codebase. Here, I'm just using this example to introduce the concept)

**Comment by fgenghini** [comment]
NIce job with the partial templates 🚀

File: assets/src/js/frontend/contribution.js [hunk 2]
```diff
@@ -1,191 +1,231 @@
...
-				}
+		// prevent negative prices
+		amount = Math.abs( amount );
 
-				// prevent non-number prices
-				if ( isNaN( price ) ) {
-					price = 0.05;
-				}
+		// correct prices outside the allowed range of 0.05 - 1000.00
+		if ( amount > 1000.0 ) {
```
Comment (sfontenot) [inline assets/src/js/frontend/contribution.js:188]: Why are purchases capped at 1000.0?

File: assets/src/js/frontend/contribution.js [hunk 3]
```diff
@@ -1,191 +1,231 @@
...
+			amount = 1000.0;
+		} else if ( amount < 0.05 ) {
+			amount = 0.05;
+		}
 
-				// prevent negative prices
-				price = Math.abs( price );
+		// format price with two digits
+		amount = amount.toFixed( 2 );
```
Comment (sfontenot) [inline assets/src/js/frontend/contribution.js:195]: Since the amount is being rounded here does it still need to be rounded on lines 174 and 176?

**Review by sfontenot — APPROVED** [review]
Added a few comments, but this lgtm :grinning:

File: inc/classes/class-client-account.php [hunk 4]
```diff
@@ -162,11 +176,14 @@ public function validate_merchant_account() {
 		$region_api_endpoints       = self::$api_endpoints[ $region ];
 		$this->connector_root       = $region_connector_endpoints['live'];
 		$this->api_root             = $region_api_endpoints['live'];
+		$this->web_endpoint         = self::$web_endpoints[ $this->merchant_region ]['live'];
 
 		// If development mode is enabled use snbox environment.
 		if ( defined( 'REVENUE_GENERATOR_ENABLE_SANDBOX' ) && true === REVENUE_GENERATOR_ENABLE_SANDBOX ) {
+			$this->is_sandbox     = true;
```
[the remark this answers is not in the document]
Comment (tomasstark) [reply inc/classes/class-client-account.php]: Great catch! This is a leftover variable that slipped my radar when making this review-ready. Removed in https://github.com/laterpay/revenue-generator/pull/277/commits/bb2ea7ec96eda580b4d9401713575efa7489e442.

File: inc/classes/class-client-account.php (hunk 1 above)
Comment (tomasstark) [reply inc/classes/class-client-account.php:351]: Excellent idea and makes total sense, @fgenghini! I refactored it in https://github.com/laterpay/revenue-generator/pull/277/commits/d17e90f8c1e72e8b3f686fc1d7794f3575a3da2b to follow the approach.

File: assets/src/js/frontend/contribution.js (hunk 3 above)
Comment (tomasstark) [reply assets/src/js/frontend/contribution.js]: Good catch! 👀 It does not! I changed this now in https://github.com/laterpay/revenue-generator/pull/277/commits/011e364f8e6a614c415f310e15bf10ae6f345713 to just use a single `.toFixed( 2 )` after lines 174 and 176.

File: assets/src/js/frontend/contribution.js (hunk 2 above)
Comment (tomasstark) [reply assets/src/js/frontend/contribution.js:190]: This was requested in June by Kim [here](https://github.com/laterpay/revenue-generator/issues/182).

**Comment by tomasstark** [comment]
@sfontenot Thanks for the review! 🙌

**Review by fgenghini — APPROVED** [review]
LGTM 🚀

**Pull request merged**